File: protocol/src/test/resources/luaTest/LuaProtocol/Packet/ObjectA.lua
-- @author jaysunxiao
-- @version 1.0
-- @since 2017 10.12 15:39

local ProtocolManager = require("LuaProtocol.ProtocolManager")

local ObjectA = {}

function ObjectA:new(a, m, objectB)
    local obj = {
        a = a, -- int
        m = m, -- java.util.Map<java.lang.Integer, java.lang.String>
        objectB = objectB -- com.zfoo.protocol.packet.ObjectB
    }
    setmetatable(obj, self)
    self.__index = self
    return obj
end

function ObjectA:protocolId()
    return 1116
end

function ObjectA:write(byteBuffer, packet)
    if packet == null then
        byteBuffer:writeBoolean(false)
        return
    end
    byteBuffer:writeBoolean(true)
    byteBuffer:writeInt(packet.a)
    if packet.m == null then
        byteBuffer:writeInt(0)
    else
        byteBuffer:writeInt(table.mapSize(packet.m))
        for key0, value1 in pairs(packet.m) do
            byteBuffer:writeInt(key0)
            byteBuffer:writeString(value1)
        end
    end
    ProtocolManager.getProtocol(1117):write(byteBuffer, packet.objectB)
end

function ObjectA:read(byteBuffer)
    if not(byteBuffer:readBoolean()) then
        return nil
    end
    local packet = ObjectA:new()
    local result2 = byteBuffer:readInt()
    packet.a = result2
    local result3 = {}
    local size4 = byteBuffer:readInt()
    if size4 > 0 then
        for index5 = 1, size4 do
            local result6 = byteBuffer:readInt()
            local result7 = byteBuffer:readString()
            result3[result6] = result7
        end
    end
    packet.m = result3
    local result8 = ProtocolManager.getProtocol(1117):read(byteBuffer)
    packet.objectB = result8
    return packet
end

return ObjectA

File: protocol/src/test/resources/jsTest/jsProtocol/packet/ObjectA.js
// @author jaysunxiao
// @version 3.0
const ObjectA = function(a, m, objectB) {
    this.a = a; // int
    this.m = m; // java.util.Map<java.lang.Integer, java.lang.String>
    this.objectB = objectB; // com.zfoo.protocol.packet.ObjectB
};

ObjectA.prototype.protocolId = function() {
    return 102;
};

ObjectA.write = function(byteBuffer, packet) {
    if (byteBuffer.writePacketFlag(packet)) {
        return;
    }
    byteBuffer.writeInt(packet.a);
    byteBuffer.writeIntStringMap(packet.m);
    byteBuffer.writePacket(packet.objectB, 103);
};

ObjectA.read = function(byteBuffer) {
    if (!byteBuffer.readBoolean()) {
        return null;
    }
    const packet = new ObjectA();
    const result0 = byteBuffer.readInt();
    packet.a = result0;
    const map1 = byteBuffer.readIntStringMap();
    packet.m = map1;
    const result2 = byteBuffer.readPacket(103);
    packet.objectB = result2;
    return packet;
};

export default ObjectA;
